### ISSUE TITLE
boinc_client does not notice (or does not care) that "computer is in use"

### ISSUE THREAD

**Issue by romw** [BOINC/boinc-ng#1187]
**Reported by Bluefin Tuna on 3 Apr 42799116 09:47 UTC**
On Linux (Fedora 17 KDE spin), boinc_client (7.0.28 x86_64-pc-linux-gnu) is running as user "boinc", a user not used for anything else.

Settings as per "Computing Preferences (apply to all BOINC projects in which you participate)" are for Processor Usage:

```
Suspend work while computer is on battery power?    yes
Matters only for portable computers 	            yes
Suspend work while computer is in use?	            yes
Suspend GPU work while computer is in use?          yes
'In use' means mouse/keyboard activity in last	    5 minutes
Suspend work if no mouse/keyboard activity in last  --- minutes
Suspend work if CPU usage is above                  50%
Do work only between the hours of                   ---
Leave tasks in memory while suspended?              no
Switch between tasks every                          120 minutes
On multiprocessors, use at most	                    --- processors
On multiprocessors, use at most                     100% of the processors
Use at most                                         75% of CPU time 
```

The machine has 2 non-threaded CPUs. As per lscpu:

```
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
CPU(s):                2
On-line CPU(s) list:   0,1
Thread(s) per core:    1
Core(s) per socket:    2
Socket(s):             1
```

It also has a CUDA-capable graphics card, which is being used for BOINC projects.

Scenario:

- boinc runs while user is absent
- once the user activates the machine, boinc notices this and tunes down
  (the log shows nothing though, I think it should ... is there a "verbose"
  option?)
  ps then shows only the "boinc_client" running, not the tasks themselves
- however, after a quarter of an hour or so, the tasks start up again; computer interactivity slows noticeably. More amazing, there are actually 3 tasks running [2 x einstein@home + 1 x seti](e.g.) even though the machine has just 2 CPUs (maybe it's done differently on CUDA machines); here too logging the decision logic would be informative
- boinc knows the machine is in use though, the log says "Suspending network activity - computer is in use"
- to get the machine in hand, one has to "kill boinc"

Migrated-From: http://boinc.berkeley.edu/trac/ticket/1215

**Comment by romw** [comment]
**Commented by ChristianB on 12 Nov 42799609 05:40 UTC**
Are you using a Fedora Package Manager installation or did you install BOINC after downloading from the official site? Packaged versions are not really supported by the official project because we don't know what the package maintainer changed and if this has some side-effects we don't know about.

To your question of a more verbose output you should look at the user wiki (http://boinc.berkeley.edu/wiki/Client_configuration) where you can find log_flags that are helpful in this case. The flags suspend_debug and task_debug may be helpful in your case.

Next question is about re-enabling computation after 45 minutes. If the user is still using the computer (mouse/keyboard activity) than this clearly is a bug. If the user is just watching a movie and not doing anything the System thinks that the computer is unused. This should be identified by the log messages. What shouldn't happen is that the computer is slowed down by the science apps. The regular BOINC client sets the process priority to lowest when starting any science application and therefore this shouldn't interfere with normal operations. Please refer to what I wrote about packaged versions of BOINC.

As GPU computing is allowed and a capable GPU is found you will always see three science processes started. One for each CPU and one for the GPU.

**Comment by romw** [comment]
**Commented by Ageless on 13 Aug 42799721 01:44 UTC**
Also know that there was this fix in 7.0.29:
- Fix for linux idle detection bug with USB mice.

For logs on what may or may not be fixed already, see [the BOINC 7 Change Log thread](http://boinc.berkeley.edu/dev/forum_thread.php?id=6698) on the BOINC forums.

**Comment by romw** [comment]
**Commented by Bluefin Tuna on 29 Oct 42800862 09:18 UTC**
Hi,

The boinc executable has been compiled from the 7.0.28 source directly.

I will add some log flags as described in the link, thank you.

In this case, I am using the computer [for editing work](e.g.) while boinc is running. This makes the slowdown very noticeable (actually, things become unacceptably slow). I don't think I saw that behaviour before I enabled the GPU, so there may be some DMA or contention problem with the graphics driver/hardware (Btw, I am using the 'nvidia' driver instead of the Fedora 'nouveau' - unsure whether I should). Will test what happens with GPU disabled.

Will also try 7.0.31, this is going to take a bit...

**Comment by romw** [comment]
**Commented by Bluefin Tuna on 15 Aug 42802452 16:43 UTC**
Update:

1) Looks like (USB) mouse mouvement is indeed not recognized. It's typing that triggers "computer in use" state.

2) The high-inertia effect is either due to the GPU or to einstein@home. Running "--no_gpus" (and thus only seti@home) triggers instant shutdown when "computer in use" is detected. If einstein@home is using the GPU, then shutdown is "rather slow" (will have to time) and for some reason response times of applications suffer a lot.

**Comment by romw** [comment]
**Commented by Bluefin Tuna on 21 Sep 42808484 17:52 UTC**
Update:

'''Without GPU'''

  * Responsiveness of PC does not suffer, irrespective of whether boinc is running or not.
  * Computer is recognized "to be in use":
    * The first time ''only'' by typing into any terminal window (not necessarily the terminal window
      belonging to the user who is running boinc). At that point boinc shutdown is rather instantaneous.
      Typing in a text editor, typing in firefox or mouse movement/clicking around is ''not' 
      recognized.
    * The second time (i.e. once boinc reactivated again because of computer idling), ''only'' by
      typing into a terminal window that was already running when boinc reactivated. A window opened
      after that is apparently not considered.
  
'''With GPU'''

  * As above, but responsiveness of PC does suffer; einstein@home must be fighting with X for GPU
    usage.

**Comment by Germano0** [comment]
Hi, I am the[ Fedora's BOINC-client co-maintainer](https://admin.fedoraproject.org/pkgdb/package/rpms/boinc-client/).
User idle time detection based on /dev/input/* does not work, since many Linux distributions like Debian, Fedora, Ubuntu do not update the "last access" of /dev/input/*
Moreover:
* you will have problems in detecting user idle time if the user is connected from SSH to a machine;
* it triggers many [SELinux denials messages](https://bugzilla.redhat.com/show_bug.cgi?id=1337607)

There is a smarter way to do that: I have seen in client/hostinfo_unix.cpp a section that concerns Apple Mac OS, you could add another section concerning [Linux distributions that use systemd](https://en.wikipedia.org/wiki/Systemd#Adoption_and_reception) and implement user idle time detection using [logind](https://www.freedesktop.org/wiki/Software/systemd/logind/) through the [D-Bus API](https://dbus.freedesktop.org/doc/api/html/index.html)

```
# gdbus introspect --system --dest org.freedesktop.login1 --object-path /org/freedesktop/login1 | grep -i idle
      readonly b IdleHint = false;
      readonly t IdleSinceHint = 0;
      readonly t IdleSinceHintMonotonic = 0;
      readonly s IdleAction = 'ignore';
      readonly t IdleActionUSec = 1800000000;
```

_Thanks to @lmctv for his suggestions_

**Comment by Germano0** [comment]
For your info, I removed the not working implementation of idle detection, from Fedora's BOINC. This avoids SELinux alerts and setroubleshot process high CPU usage

**Comment by ChristianBeer** [comment]
I did a bit experimenting with the sd-bus API to determine the idle state. I have a small test program but it always reports the system as not idle. I also found that the IdleHint is only changed when the Desktop Manager (like Gnome and KDE) reports the system as idle. So this would not work for systems that don't have a Desktop Manager. I still think it's worth to investigate more so it at least works for systems with a DM. I'll keep trying.

**Comment by Germano0** [comment]
Hi @ChristianBeer , systemd's logind idle detection should really work even without a running Desktop Manager

**Comment by Germano0** [comment]
I have seen changelog
https://boinc.berkeley.edu/dev/forum_thread.php?id=10204&postid=67990
and I was wondering about the state of this bugreport. Unfortunately I still had not the opportunity to make some tests due
https://boinc.berkeley.edu/dev/forum_thread.php?id=11446

**Comment by Germano0** [comment]
I made a dff between `hostinfo_unix.cpp` of boinc 7.6.22 and 7.6.33:

```
$ diff boinc_7.6.22_hostinfo_unix.cpp boinc_7.6.33_hostinfo_unix.cpp 
43a44,53
> #if HAVE_XSS
> #include <X11/extensions/scrnsaver.h> //X-based idle detection
> // prevents naming collision between X.h define of Always and boinc's
> // lib/prefs.h definition in an enum.
> #undef Always
> #include <dirent.h> //for opening /tmp/.X11-unix/
>   // (There is a DirScanner class in BOINC, but it doesn't do what we want)
> #include "log_flags.h" // idle_detection_debug flag for verbose output
> #endif
> 
1314,1315c1324,1325
<     strncpy( p_model, "Alpha ", sizeof( p_model));
<     strncat( p_model, cpu_type_name, (sizeof( p_model)- strlen( p_model)- 1));
---
>     strlcpy(p_model, "Alpha ", sizeof(p_model));
>     strlcat(p_model, cpu_type_name, sizeof(p_model));
1949a1960,2146
> 
> #if HAVE_XSS
> 
> // Initializer for const vector<string> in xss_idle
> //
> const vector<string> X_display_values_initialize() {
> 
>   /* According to "man Xserver", each local Xserver will have a socket file
>    * at /tmp/.X11-unix/Xn, where "n" is the display number (0, 1, 2, etc).
>    * We will parse this directory for currently open Xservers and attempt
>    * to ultimately query them for their idle time. If we can't open this
>    * directory, or the display_values vector is otherwise empty, then a
>    * static list of guesses for open display servers is utilized instead
>    * (DISPLAY values ":{0..6}") that will attempt connections to the first
>    * seven open Xservers.
>    *
>    * If we were unable to open _any_ Xserver, then we will log this and
>    * xss_idle returns true, effectively leaving idle detection up to other
>    * methods.
>    */
>   static const string dir = "/tmp/.X11-unix/";
>   vector<string> display_values;
>   vector<string>::iterator it;
> 
>   DIR *dp;
>   struct dirent *dirp;
>   if((dp = opendir(dir.c_str())) == NULL) {
>     if ( log_flags.idle_detection_debug ) {
>       msg_printf(NULL, MSG_INFO, 
>         "[idle_detection] Error (%d) opening %s.", errno, dir.c_str());
>     }
>   }
> 
>   while ((dirp = readdir(dp)) != NULL) {
>     display_values.push_back(string(dirp->d_name));
>   }
>   closedir(dp);
> 
>   // Get rid of non-matching elements and format the matching ones.
>   for ( it = display_values.begin() ; it != display_values.end() ; ) {
>     if ( it->c_str()[0] != 'X' ) {
>       it = display_values.erase(it);
>     }
>     else {
>       replace(it->begin(), it->end(), 'X', ':');
>       it++;
>     }
> 
>   }
> 
>   // if the display_values vector is empty, assume something went wrong
>   // (couldn't open directory, no apparent Xn files). Test a static list of
>   // DISPLAY values instead that is likely to catch most common use cases.
>   // (I don't know of many environments where there will simultaneously be
>   // more than seven active, local Xservers. I'm sure they exist... somewhere.
>   // But seven was the magic number for me).
>   if ( display_values.size() == 0 ) {
>     if ( log_flags.idle_detection_debug ) {
>       msg_printf(NULL, MSG_INFO,
>         "[idle_detection] No DISPLAY values found in /tmp/.X11-unix/.");
>       msg_printf(NULL, MSG_INFO,
>         "[idle_detection] Using static DISPLAY list, :{0..6}.");
>     }
>     display_values.push_back(":0");
>     display_values.push_back(":1");
>     display_values.push_back(":2");
>     display_values.push_back(":3");
>     display_values.push_back(":4");
>     display_values.push_back(":5");
>     display_values.push_back(":6");
>     return display_values;
>   }
>   else {
>     return display_values;
>   }
> }
> 
> /* Ask the X server for user idle time (using XScreenSaver API)
>  * Return true if the idle time exceeds idle_threshold for all accessible
>  * Xservers. However, if even one Xserver reports busy/non-idle, then
>  * return false. This function assumes that the boinc user has been
>  * granted access to the Xservers a la "xhost +SI:localuser:boinc". If
>  * access isn't available for an Xserver, then that Xserver is skipped.
>  * One may drop a file in /etc/X11/Xsession.d/ that runs the xhost command
>  * for all Xservers on a machine when the Xservers start up.
>  */
> bool xss_idle(long idle_threshold) {
> 
>   const vector<string> display_values = X_display_values_initialize();
>   vector<string>::const_iterator it;
>   // If we can connect to at least one DISPLAY, this is set to false.
>   bool no_available_x_display = true;
> 
>   static XScreenSaverInfo* xssInfo = XScreenSaverAllocInfo();
>   // This shouldn't fail. XScreenSaverAllocInfo just returns a small
>   // struct (see "man 3 xss"). If we can't allocate this, then we've
>   // got bigger problems to worry about.
>   if ( xssInfo == NULL ) {
>     if ( log_flags.idle_detection_debug ) {
>       msg_printf(NULL, MSG_INFO,
>         "[idle_detection] XScreenSaverAllocInfo failed. Out of memory? Skipping XScreenSaver idle detection.");
>     }
>     return true;
>   }
> 
>   for (it = display_values.begin() ; it != display_values.end() ; it++) {
> 
>     Display* disp = NULL;
>     long idle_time = 0;
>     
>     disp = XOpenDisplay(it->c_str());
>     // XOpenDisplay may return NULL if there is no running X
>     // or DISPLAY points to wrong/invalid display
>     if (disp == NULL) {
>       if ( log_flags.idle_detection_debug ) {
>       msg_printf(NULL, MSG_INFO, 
>       "[idle_detection] DISPLAY '%s' not found or insufficient access.",
>       it->c_str());
>       }
>       continue;
>     }
> 
>     // Determine if the DISPLAY we have accessed has the XScreenSaver
>     // extension or not.
>     int event_base_return, error_base_return;
>     if (!XScreenSaverQueryExtension(
>       disp, &event_base_return, &error_base_return
>     )){
>       if ( log_flags.idle_detection_debug ) {
>       msg_printf(NULL, MSG_INFO,
>         "[idle_detection] XScreenSaver extension not available for DISPLAY '%s'.",
>         it->c_str());
>       }
>       XCloseDisplay(disp);
>       continue;
>     }
> 
>     // All checks passed. Get the idle information.
>     no_available_x_display = false;
>     XScreenSaverQueryInfo(disp, DefaultRootWindow(disp), xssInfo);
>     idle_time = xssInfo->idle;
> 
>     // Close the connection to the XServer
>     XCloseDisplay(disp);
> 
>     // convert from milliseconds to seconds
>     idle_time = idle_time / 1000;
> 
>     if ( log_flags.idle_detection_debug ) {
>       msg_printf(NULL, MSG_INFO, 
>         "[idle_detection] XSS idle detection succeeded on DISPLAY '%s'.", it->c_str());
>       msg_printf(NULL, MSG_INFO, 
>         "[idle_detection] idle threshold: %ld", idle_threshold);
>       msg_printf(NULL, MSG_INFO,
>         "[idle_detection] idle_time: %ld", idle_time);
>     }
> 
>     if ( idle_threshold < idle_time ) {
>       if ( log_flags.idle_detection_debug ) {
>         msg_printf(NULL, MSG_INFO,
>           "[idle_detection] DISPLAY '%s' is idle.", it->c_str());
>       }
>     } else {
>       if ( log_flags.idle_detection_debug ) {
>         msg_printf(NULL, MSG_INFO,
>           "[idle_detection] DISPLAY '%s' is active.", it->c_str());
>       }
>       return false;
>     }
> 
>   }
> 
>   /* We should only ever get here if all Xservers (that were queryable) were
>    * idle. If none of the Xservers were queryable, we should still end up here,
>    * and simply report true. In that case, the xss_idle function effectively
>    * provides no information on the idle state of the system, as no Xservers
>    * were accessible to interrogate.
>    */
>   if ( log_flags.idle_detection_debug && no_available_x_display ) {
>     msg_printf(NULL, MSG_INFO,
>       "[idle_detection] Could not connect to any DISPLAYs. XSS idle determination impossible.");
>   }
>   return true;
> 
> }
> #endif // HAVE_XSS
> 
1974a2172,2177
> #if HAVE_XSS
>     if (!xss_idle((long)(idle_time_to_run * 60))) {
>         return false;
>     }
> #endif // HAVE_XSS
> 
1993c2196
< #endif
---
> #endif // LINUX_LIKE_SYSTEM
```

So it looks like to now rely on X server to detect user idle time. I would like to ask you if you are interested in dropping the old code that runs `stat` on /dev/input and that now is used as "fallback" in case of failure of X detection, or you are interested in keeping it.
Actually I just started testing BOINC 7.6.33 on Fedora 25

**Comment by JuhaSointusalo** [comment]
> So it looks like that now BOINC relies on X server to detect user idle time.

Please notice that BOINC client needs permission to access X server or the XSS based idle detection won't work. See the discussion in #1453. Debian uses [this file](https://anonscm.debian.org/cgit/pkg-boinc/boinc.git/tree/debian/36x11-common_xhost-boinc) in etc/X11/Xsession.d/ .

> interested in dropping the old code that runs stat on /dev/input

Maybe check with @romw what distro he used when he wrote the old code and if it ever worked.
And check a variety of distros of different ages and make sure it doesn't work anywhere.
And maybe check kernel build options that there isn't any switch that makes the old code work.
And then throw the old code away.

**Comment by LocutusOfBorg** [comment]
Mmm what about headless servers? AFAIR this code was meant to work with them

**Comment by ChristianBeer** [comment]
We should definitely keep the fallback for systems that don't have X. What @Germano0 wants is a way to disable the access to `/dev` for certain Linux distributions that use SELinux and where this is triggering an alarm. Instead of removing the code we could make it possible to disable it at compile time and show a message at startup that idle detection does not work without an X server with this package.

**Comment by JuhaSointusalo** [comment]
@LocutusOfBorg @ChristianBeer 

It's the `all_input_idle()` function we are talking about. It tries to detect activity by checking access time of `event` and `mouse` devices in `/dev/input` . @Germano0 said it doesn't work on Debian, Ubuntu and Fedora and I just tested that it doesn't work on Mint 17 = Ubuntu 14.04 kernel 3.13.0-105-generic and Rasbian Wheezy kernel 3.18.16+. The problem is that the access time is updated only when the device node is created, that is, when you plug in the device.

If it turns out that there is not a single distro where the code works then why bother keep it.

There is a similar test for pseudoterminal devices though you need to be using terminal for the test to work.

**Comment by Germano0** [comment]
> Instead of removing the code we could make it possible to disable it at compile time and show a message at startup that idle detection does not work without an X server with this package.

It would be great to have two different flags to disable:
- /dev/input based idle detection;
- XScreenSaver API based idle detection.

**Comment by Germano0** [comment]
On Fedora we will patch BOINC 7.6.33 to remove Xorg based idle detection too, because Fedora is using Wayland by default and Fedora servers do not run graphical server as @LocutusOfBorg said.

**Comment by ChristianBeer** [comment]
Summary of discussion above:
Some Linux distributions don't want the BOINC Client to access anything in `/dev` or there is no XServer available at runtime. Thus the conclusion was to provide two build time flags to disable:
- /dev/input based idle detection;
- XScreenSaver API based idle detection.

The default should remain to be both enabled.

**Comment by Ageless93** [comment]
@lfield, can you check if this is still a problem or requirement?

**Comment by Germano0** [comment]
> @lfield, can you check if this is still a problem or requirement?

@Ageless93 what do you mean?

**Comment by Ageless93** [comment]
Laurence Field (lfiend) is the Linux BOINC Release Manager. I'm pointing out issues that he may want to have fixed or check that they are fixed already so these issues can be closed.

**Comment by Germano0** [comment]
@Ageless93  As Fedora BOINC co-maintainer together with @lfield I can say that actual situation is:
- `/dev/input*` stat based idle time detection: code is still there, it still does not work;
- "90s" xorg based idle time detection: code is still there, it still does not work because it thinks the system is always in use, even if your idle time reachs the time in user settings;
- still no feedback about https://github.com/vorot93/volunode-old/issues/5
- I haven't seen any flag to disable idle time detection as stated in https://github.com/BOINC/boinc/issues/1187#issuecomment-292879132

**Comment by lfield** [comment]
Having thought about this, I think that we should define the 'in use' option to only be relevant when there is a window manager. Having said that, it looks like this option is still not working.

**Comment by kgerickson** [comment]
This is still a problem.  Would something like this work as a solution:

https://www.freshports.org/x11/xidle/

**Comment by Germano0** [comment]
@kgerickson no https://github.com/BOINC/boinc/issues/1187#issuecomment-339070513

**Comment by RubenKelevra** [comment]
That's still an issue, I'm running:

- Linux 4.16.13
- GDM 3.28.1
- Gnome 3.28.2
- Boinc 7.8.6
- seti@home

Sometimes the keyboard/mouse activity is detected, and BOINC pauses, just to continue the minute after, or maybe 2-3 minutes later. It looks not completely broken, but not reliable detecting the activity. 

It's not fixed for 3 years now? come on guys!

**Comment by laggarcia** [comment]
This is still an issue in Fedora 29 :(

**Comment by BryanQuigley** [comment]
Has anyone found a systemd/dbus way to do this, or should we just start doing desktop environments?

**Comment by lfield** [comment]
For me idle detection is a desktop environment feature. CPU load is the default metric otherwise.

**Comment by Germano0** [comment]
> Has anyone found a systemd/dbus way to do this

Among my other activities, I am slowly playing with GDbus API of Glib/GIO. I said "slowly" because the documentation is very "sterile", and not helpful at all. Luckly there are some softwares that uses them, so you can take examples (like darktable)

> or should we just start doing desktop environments?

????

**Comment by laggarcia** [comment]
dbus-send --print-reply --dest=org.gnome.Mutter.IdleMonitor /org/gnome/Mutter/IdleMonitor/Core org.gnome.Mutter.IdleMonitor.GetIdletime

The above call should give you idle time on gnome-shell (GNOME3) environments on top of Wayland. Not a perfect solution, but probably better than what we have today at least for Fedora default installation.

**Comment by laggarcia** [comment]
Another very interesting way to detect idle time for any user logged into a Linux system is to use the same approach used by the w tool. It basically get stat for the current terminal file (graphical, text, or serial) and watch for the last access timestamp (stat.st_atime).

**Comment by Germano0** [comment]
The following comment illustrates how I was trying to resolve this problem, and the conclusion I arrived to.

Unfortunately we cannot use systemd-logind to detect user idle time because it relies on desktop environments to pass this information, but none does it. So I was starting to implement it for for GNOME and KDE Plasma. For the former, there is a DBus solution, for the latter instead, KIdleTime API is needed.

I managed  [to implement a raw prototype](https://github.com/BOINC/boinc/files/3262003/gnome_idle.txt) of BOINC user idle time detection for GNOME, and it works if the client is runned as user process. It cannot work with BOINC client running as a service because it runs as unprivileged and confined user, instead GNOME exposes GetIdletime method on DBus user session that is not reachable for processes that are not running under the same user.

In my opinion a good solution can be the following
![new BOINC client prototype](https://user-images.githubusercontent.com/5477747/59039362-fcc71480-8874-11e9-89fa-4205d5fbac77.jpg)
[vectorial version of the image](https://github.com/BOINC/boinc/files/3261731/new.BOINC.client.prototype.pdf)


BOINC client will continue running as confined service, but we implement some code to let it be on DBus *system* session.
We implement a new very tiny process that automatically runs on user space (thanks to a user space systemd unit file in path /etc/systemd/user). Since it will run with user privileges, it can talk to things that are running as user level, like objects that are connected to DBus *user* session as GNOME Idle Monitor, or directly to KDE IdleTime API. Moreover the tiny process will talk to BOINC client by connecting also to DBus *system* session.

What do you think about this proposal?
Is it worth implementing, or are we introducing too much complexity for the Linux client? 

**Bibliography**
[[systemd-devel] interacting with logind to detect user idle time](https://lists.freedesktop.org/archives/systemd-devel/2019-May/042496.html)
[[GNOME discourse] IdleMonitor (DBus) and unprivileged user](https://discourse.gnome.org/t/idlemonitor-dbus-and-unprivileged-user/1120)
[[KDE Plasma devel] Concerning KIdleTime API (was Proposal: passing user idle time to systemd-logind)](https://mail.kde.org/pipermail/plasma-devel/2019-May/098194.html)
Concerning systemd user services Tom Huges and grift on #fedora-devel Freenode IRC channel

**Comment by BryanQuigley** [comment]
Thanks for running these down!  I actually just asked systemd about doing a more kidletime like implementation in https://github.com/systemd/systemd/issues/12668  (although I didn't know what I was asking for was like kidletime until reading the thread you started)

If we can't get it in systemd proper (or somewhere else universal) I think we do need another simple utility unfortunately.  It definitely should be handled by something on the system and not BOINC, IMHO (no pull with project though).  I'll review it more when I have time.

**Comment by Germano0** [comment]
@BryanQuigley you may want to attach your systemd ticket URL with a comment, to the systemd mailing list thread I created about user idle time
>  [[systemd-devel] interacting with logind to detect user idle time](https://lists.freedesktop.org/archives/systemd-devel/2019-May/042496.html)

**Comment by Fifis** [comment]
Still does not work properly in Arch Linux on BOINC 7.14.2.

**Comment by lockywolf** [comment]
Still doesn't work properly on Slackware-current 2020-03-28 when run as a different (dedicated) user. 

Perhaps some documentation on how to permit x11 connection or something like that would be nice to have easily accessible in the gui.

**Comment by jigpu** [comment]
> BOINC client will continue running as confined service, but we implement some code to let it be on DBus _system_ session.
> We implement a new very tiny process that automatically runs on user space (thanks to a user space systemd unit file in path /etc/systemd/user). Since it will run with user privileges, it can talk to things that are running as user level, like objects that are connected to DBus _user_ session as GNOME Idle Monitor, or directly to KDE IdleTime API. Moreover the tiny process will talk to BOINC client by connecting also to DBus _system_ session.
> 
> What do you think about this proposal?
> Is it worth implementing, or are we introducing too much complexity for the Linux client?
> 
Looks reasonable, though I don't think there's any particular need to deal with the system dbus. BOINC already has its own RPC mechanism that could be extended to include an "idle" interface. Your client (which could be a small independent program, or even the BOINC GUI) would periodically forward idle information from the session dbus through RPC and then BOINC would decide if enough idle time has passed. There would probably need to be a little care put into handling multiple simultaneous sessions (e.g. I've switched away from my roomates session, so BOINC sees both near-zero idle times from my session client, and very high idle times for the other session's client), but simply using the minimum of recent reports should be fine.

**Comment by evilbenfranklin** [comment]
Still an issue on Linux 5.4, Manjaro, using KDE Plasma and BOINC 7.17.0 built from the AUR. The client always thinks the system is in use if any of the "in use" options are enabled.

**Comment by LinAGKar** [comment]
For me, on OpenSUSE Tumbleweed KDE, boinc always acts like the system is idle. It used to work, at least if I did `xhost si:localuser:boinc`, but it got broken a few years ago.

**Comment by jigpu** [comment]
Just noticed this commit that looks like it might fix this issue: https://github.com/BOINC/boinc/commit/8a1099f816652cffdd24d8d0c91b36e6ae7ef031

**Comment by LinAGKar** [comment]
Not my issue at least. The comment there says that problem started in 7.16.4, but I'm on 7.16.3.

**Comment by BryanQuigley** [comment]
There are multiple issues, but if running a modern version it might be X detection is blocked (found during #3549 testing).  Try editing boinc service so PrivateTmp=false to re-enable real X detection.

**Comment by evilbenfranklin** [comment]
> There are multiple issues, but if running a modern version it might be X detection is blocked (found during #3549 testing). Try editing boinc service so PrivateTmp=false to re-enable real X detection.

Unfortunately that did not work for me. Current behavior is as follows under these specs: 
```
OS: Manjaro Linux x86_64 
Host: MS-7A38 2.0 
Kernel: 5.4.33-3-MANJARO 
Uptime: 1 day, 17 hours, 5 mins 
Packages: 2429 (pacman), 1 (flatpak) 
Shell: zsh 5.8 
Resolution: 1600x900, 1920x1080, 1920x1080 
DE: Plasma 
WM: KWin 
WM Theme: Breeze 
Theme: Breath2 [Plasma], Breath-Dark [GTK2/3] 
Icons: breath2 [Plasma], breath2 [GTK2/3] 
Terminal: konsole 
CPU: AMD Ryzen 5 1600 (12) @ 3.400GHz 
GPU: NVIDIA GeForce GTX 1060 6GB 
Memory: 13547MiB / 32137MiB 
```

If compute preferences are set to suspend while computer is in use, it constantly stays suspended, but if set to suspend only GPU while is in use or when CPU usage is about 25%, it will run as expected, just without the GPU component.

If I let it go idle under those settings (GPU/CPU > 25%), then GPU still remains suspended.

**Comment by jigpu** [comment]
> There are multiple issues, but if running a modern version it might be X detection is blocked (found during #3549 testing). Try editing boinc service so PrivateTmp=false to re-enable real X detection.

Also did not work for me on Arch Linux / GNOME 3.36. Compiling 7.16.6 with commit 8a1099f patched in seems to have done the trick, however (and also didn't require a manual edit to `PrivateTmp`).

**Comment by LinAGKar** [comment]
> There are multiple issues, but if running a modern version it might be X detection is blocked (found during #3549 testing). Try editing boinc service so PrivateTmp=false to re-enable real X detection

That doesn't seem to be it either. I didn't have PrivateTmp there, and adding it didn't help.

It's acting really wonky. It did suspend work on the GPU now, but only after I'd been using the computer for like half a minute, and then it started work again a few minutes later.

I tried enabling idle_detection_debug, but nothing shows up in the log.

**Comment by LinAGKar** [comment]
It looks like someone has deliberately broken idle detection: https://build.opensuse.org/request/show/627693

**Comment by debianmain1** [comment]
I am also having problems with GPUs not starting after selecting "suspend GPU computing when computer is in use" Using current Debian Sid & Boinc Manager 7.16.6 from Debian.
 Tried the "PrivateTmp=false" without success. Anyone have another idea?

**Comment by BryanQuigley** [comment]
I just ran into /run/systemd/[session/user/seat].  It does actually seem to indicate when a session is active or not in xfce.  STATE=active -> STATE=online.   Need to investigate further...

**Comment by lockywolf** [comment]
Does elogind have the same feature?

**Comment by aaronpuchert** [comment]
@Germano0's proposals seems to be proper architecture for this. Idle detection needs to happen outside the client in the user session, so we need a separate service for it.

> BOINC client will continue running as confined service, but we implement some code to let it be on DBus _system_ session.
> We implement a new very tiny process that automatically runs on user space (thanks to a user space systemd unit file in path /etc/systemd/user). Since it will run with user privileges, it can talk to things that are running as user level, like objects that are connected to DBus _user_ session as GNOME Idle Monitor, or directly to KDE IdleTime API. Moreover the tiny process will talk to BOINC client by connecting also to DBus _system_ session.

I don't think we need the user space daemon to listen on the DBus system session, it could just use the existing BOINC RPC to communicate with the client, e.g. run `boinccmd --set_run_mode auto` when the session is idle and `boinccmd --set_run_mode never` when the session is active.

So it would only need to be on the user DBus, or for a prototype perhaps just run `loginctl show-session`. It could even be a script.

**Comment by spark-corona** [comment]
Got the same problem here.

System specs are:

boinc --version: 7.16.6 x86_64-pc-linux-gnu

uname -a: Linux mobile-workstation 5.13.0-27-generic #29~20.04.1-Ubuntu SMP Fri Jan 14 00:32:30 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

lshw: 

    Beschreibung: Notebook
    Produkt: XMG_APEX15_XAP15E20
    Hersteller: SchenkerTechnologiesGmbH
    Version: Not Applicable
    Seriennummer: Not Applicable
    Breite: 64 bits
    Fähigkeiten: smbios-3.2.0 dmi-3.2.0 smp vsyscall32
    Konfiguration: administrator_password=disabled boot=normal chassis=notebook family=Not Applicable power-on_password=disabled uuid=000096F0-0000-0010-A100-80FA5B8BA443
  *-core
       Beschreibung: Hauptplatine
       Produkt: NH5xAx
       Hersteller: NB01
       Physische ID: 0
       Version: Not Applicable
       Seriennummer: Not Applicable
       Steckplatz: To Be Filled By O.E.M.
     *-firmware
          Beschreibung: BIOS
          Hersteller: Insyde Corp.
          Physische ID: 0
          Version: 1.07.07RTR1
          date: 04/08/2020
          Größe: 128KiB
          Kapazität: 6784KiB
          Fähigkeiten: pci upgrade shadowing cdboot bootselect socketedrom edd int13floppynec int13floppytoshiba int13floppy360 int13floppy1200 int13floppy720 int13floppy2880 int9keyboard int10video acpi usb uefi
     *-cpu
          Beschreibung: CPU
          Produkt: AMD Ryzen 9 PRO 3900 12-Core Processor
          Hersteller: Advanced Micro Devices [AMD]
          Physische ID: 4
          Bus-Informationen: cpu@0
          Version: AMD Ryzen 9 PRO 3900 12-Core Processor
          Seriennummer: Unknown
          Steckplatz: AM4
          Größe: 3770MHz
          Kapazität: 4359MHz
          Breite: 64 bits
          Takt: 100MHz
          Fähigkeiten: lm fpu fpu_exception wp vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp x86-64 constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es cpufreq
          Konfiguration: cores=12 enabledcores=12 threads=24
        *-cache:0
             Beschreibung: L1 Cache
             Physische ID: 5
             Steckplatz: L1 - Cache
             Größe: 768KiB
             Kapazität: 768KiB
             Takt: 1GHz (1.0ns)
             Fähigkeiten: pipeline-burst internal write-back unified
             Konfiguration: level=1
        *-cache:1
             Beschreibung: L2 Cache
             Physische ID: 6
             Steckplatz: L2 - Cache
             Größe: 6MiB
             Kapazität: 6MiB
             Takt: 1GHz (1.0ns)
             Fähigkeiten: pipeline-burst internal write-back unified
             Konfiguration: level=2
        *-cache:2
             Beschreibung: L3 Cache
             Physische ID: 7
             Steckplatz: L3 - Cache
             Größe: 64MiB
             Kapazität: 64MiB
             Takt: 1GHz (1.0ns)
             Fähigkeiten: pipeline-burst internal write-back unified
             Konfiguration: level=3
     *-memory
          Beschreibung: Systemspeicher
          Physische ID: 25
          Steckplatz: Systemplatine oder Hauptplatine
          Größe: 32GiB
        *-bank:0
             Beschreibung: SODIMM DDR4 Synchron Unbuffered (Unregistered) 2667 MHz (0,4 ns)
             Produkt: M471A2K43DB1-CTD
             Hersteller: Samsung
             Physische ID: 0
             Seriennummer: 40B3B190
             Steckplatz: DIMM 0
             Größe: 16GiB
             Breite: 64 bits
             Takt: 2667MHz (0.4ns)
        *-bank:1
             Beschreibung: SODIMM DDR4 Synchron Unbuffered (Unregistered) 2667 MHz (0,4 ns)
             Produkt: M471A2K43DB1-CTD
             Hersteller: Samsung
             Physische ID: 1
             Seriennummer: 40B3B1C0
             Steckplatz: DIMM 0
             Größe: 16GiB
             Breite: 64 bits
             Takt: 2667MHz (0.4ns)
     *-pci:0
          Beschreibung: Host bridge
          Produkt: Starship/Matisse Root Complex
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 100
          Bus-Informationen: pci@0000:00:00.0
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
        *-generic UNGEFORDERT
             Beschreibung: IOMMU
             Produkt: Starship/Matisse IOMMU
             Hersteller: Advanced Micro Devices, Inc. [AMD]
             Physische ID: 0.2
             Bus-Informationen: pci@0000:00:00.2
             Version: 00
             Breite: 32 bits
             Takt: 33MHz
             Fähigkeiten: msi ht cap_list
             Konfiguration: latency=0
        *-pci:0
             Beschreibung: PCI bridge
             Produkt: Starship/Matisse GPP Bridge
             Hersteller: Advanced Micro Devices, Inc. [AMD]
             Physische ID: 1.1
             Bus-Informationen: pci@0000:00:01.1
             Version: 00
             Breite: 32 bits
             Takt: 33MHz
             Fähigkeiten: pci pm pciexpress msi ht normal_decode bus_master cap_list
             Konfiguration: driver=pcieport
             Ressourcen: irq:26 memory:f2900000-f29fffff
           *-storage
                Beschreibung: Non-Volatile memory controller
                Produkt: Samsung Electronics Co Ltd
                Hersteller: Samsung Electronics Co Ltd
                Physische ID: 0
                Bus-Informationen: pci@0000:01:00.0
                Version: 00
                Breite: 64 bits
                Takt: 33MHz
                Fähigkeiten: storage pm msi pciexpress msix nvm_express bus_master cap_list
                Konfiguration: driver=nvme latency=0
                Ressourcen: irq:44 memory:f2900000-f2903fff
              *-nvme0
                   Beschreibung: NVMe device
                   Produkt: SAMSUNG MZVLQ1T0HALB-00000
                   Physische ID: 0
                   Logischer Name: /dev/nvme0
                   Version: FXV7101Q
                   Seriennummer: S4Y8NF0N700510
                   Konfiguration: nqn=nqn.1994-11.com.samsung:nvme:PM991:M.2:S4Y8NF0N700510 state=live
                 *-namespace
                      Beschreibung: NVMe namespace
                      Physische ID: 1
                      Logischer Name: /dev/nvme0n1
                      Größe: 953GiB (1024GB)
                      Fähigkeiten: gpt-1.00 partitioned partitioned:gpt
                      Konfiguration: guid=9cdff522-f0ee-482b-af09-30c79b5fa165 logicalsectorsize=512 sectorsize=512
                    *-volume:0
                         Beschreibung: Windows FAT Laufwerk
                         Hersteller: MSDOS5.0
                         Physische ID: 1
                         Logischer Name: /dev/nvme0n1p1
                         Logischer Name: /boot/efi
                         Version: FAT32
                         Seriennummer: 1e18-5893
                         Größe: 1497MiB
                         Kapazität: 1499MiB
                         Fähigkeiten: boot fat initialized
                         Konfiguration: FATs=2 filesystem=fat label=BOOT mount.fstype=vfat mount.options=rw,relatime,fmask=0077,dmask=0077,codepage=437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro name=EFI system partition state=mounted
                    *-volume:1
                         Beschreibung: reserved partition
                         Hersteller: Windows
                         Physische ID: 2
                         Logischer Name: /dev/nvme0n1p2
                         Seriennummer: aac8f6d1-2806-4112-a18d-85a3fc194e37
                         Kapazität: 1999MiB
                         Fähigkeiten: nofs
                         Konfiguration: name=Microsoft reserved partition
                    *-volume:2
                         Beschreibung: Windows NTFS Laufwerk
                         Hersteller: Windows
                         Physische ID: 3
                         Logischer Name: /dev/nvme0n1p3
                         Version: 3.1
                         Seriennummer: 3a18-ec84
                         Größe: 1986MiB
                         Kapazität: 1999MiB
                         Fähigkeiten: boot ntfs initialized
                         Konfiguration: clustersize=4096 created=2020-03-22 10:34:34 filesystem=ntfs label=Recovery name=Basic data partition state=clean
                    *-volume:3
                         Beschreibung: Windows NTFS Laufwerk
                         Hersteller: Windows
                         Physische ID: 4
                         Logischer Name: /dev/nvme0n1p4
                         Version: 3.1
                         Seriennummer: 8e2c109a-37f8-6240-ac66-ad2b161ab647
                         Größe: 148GiB
                         Kapazität: 148GiB
                         Fähigkeiten: ntfs initialized
                         Konfiguration: clustersize=4096 created=2020-03-22 10:34:38 filesystem=ntfs label=Windows name=Basic data partition state=dirty
                    *-volume:4
                         Beschreibung: EXT4-Laufwerk
                         Hersteller: Linux
                         Physische ID: 5
                         Logischer Name: /dev/nvme0n1p5
                         Logischer Name: /
                         Version: 1.0
                         Seriennummer: 8804817e-d4d6-4a3e-827e-c90f97c0d3c8
                         Größe: 799GiB
                         Fähigkeiten: journaled extended_attributes large_files huge_files dir_nlink recover 64bit extents ext4 ext2 initialized
                         Konfiguration: created=2022-01-04 12:25:16 filesystem=ext4 lastmountpoint=/ modified=2022-01-31 17:10:07 mount.fstype=ext4 mount.options=rw,relatime,errors=remount-ro mounted=2022-01-31 17:10:08 state=mounted
        *-pci:1
             Beschreibung: PCI bridge
             Produkt: Starship/Matisse GPP Bridge
             Hersteller: Advanced Micro Devices, Inc. [AMD]
             Physische ID: 1.3
             Bus-Informationen: pci@0000:00:01.3
             Version: 00
             Breite: 32 bits
             Takt: 33MHz
             Fähigkeiten: pci pm pciexpress msi ht normal_decode bus_master cap_list
             Konfiguration: driver=pcieport
             Ressourcen: irq:27 ioport:3000(Größe=4096) memory:f2600000-f28fffff
           *-usb
                Beschreibung: USB controller
                Produkt: 400 Series Chipset USB 3.1 XHCI Controller
                Hersteller: Advanced Micro Devices, Inc. [AMD]
                Physische ID: 0
                Bus-Informationen: pci@0000:02:00.0
                Version: 01
                Breite: 64 bits
                Takt: 33MHz
                Fähigkeiten: msi pm pciexpress xhci bus_master cap_list
                Konfiguration: driver=xhci_hcd latency=0
                Ressourcen: irq:82 memory:f2820000-f2827fff
              *-usbhost:0
                   Produkt: xHCI Host Controller
                   Hersteller: Linux 5.13.0-27-generic xhci-hcd
                   Physische ID: 0
                   Bus-Informationen: usb@1
                   Logischer Name: usb1
                   Version: 5.13
                   Fähigkeiten: usb-2.00
                   Konfiguration: driver=hub slots=10 speed=480Mbit/s
                 *-usb:0
                      Beschreibung: Tastatur
                      Produkt: USB Receiver
                      Hersteller: Logitech
                      Physische ID: 2
                      Bus-Informationen: usb@1:2
                      Version: 12.11
                      Fähigkeiten: usb-2.00
                      Konfiguration: driver=usbhid maxpower=98mA speed=12Mbit/s
                 *-usb:1 UNGEFORDERT
                      Beschreibung: Allgemeines USB-Gerät
                      Hersteller: Synaptics, Inc.
                      Physische ID: 7
                      Bus-Informationen: usb@1:7
                      Version: 1.54
                      Seriennummer: d4059c0024f0
                      Fähigkeiten: usb-2.00
                      Konfiguration: maxpower=100mA speed=12Mbit/s
                 *-usb:2
                      Beschreibung: Bluetooth-Schnittstelle
                      Hersteller: Intel Corp.
                      Physische ID: a
                      Bus-Informationen: usb@1:a
                      Version: 0.01
                      Fähigkeiten: bluetooth usb-2.01
                      Konfiguration: driver=btusb maxpower=100mA speed=12Mbit/s
              *-usbhost:1
                   Produkt: xHCI Host Controller
                   Hersteller: Linux 5.13.0-27-generic xhci-hcd
                   Physische ID: 1
                   Bus-Informationen: usb@2
                   Logischer Name: usb2
                   Version: 5.13
                   Fähigkeiten: usb-3.10
                   Konfiguration: driver=hub slots=4 speed=10000Mbit/s
           *-sata
                Beschreibung: SATA controller
                Produkt: 400 Series Chipset SATA Controller
                Hersteller: Advanced Micro Devices, Inc. [AMD]
                Physische ID: 0.1
                Bus-Informationen: pci@0000:02:00.1
                Version: 01
                Breite: 32 bits
                Takt: 33MHz
                Fähigkeiten: sata msi pm pciexpress ahci_1.0 bus_master cap_list rom
                Konfiguration: driver=ahci latency=0
                Ressourcen: irq:43 memory:f2800000-f281ffff memory:f2880000-f28fffff
           *-pci
                Beschreibung: PCI bridge
                Produkt: 400 Series Chipset PCIe Bridge
                Hersteller: Advanced Micro Devices, Inc. [AMD]
                Physische ID: 0.2
                Bus-Informationen: pci@0000:02:00.2
                Version: 01
                Breite: 32 bits
                Takt: 33MHz
                Fähigkeiten: pci msi pm pciexpress normal_decode bus_master cap_list
                Konfiguration: driver=pcieport
                Ressourcen: irq:34 ioport:3000(Größe=4096) memory:f2600000-f27fffff
              *-pci:0
                   Beschreibung: PCI bridge
                   Produkt: 400 Series Chipset PCIe Port
                   Hersteller: Advanced Micro Devices, Inc. [AMD]
                   Physische ID: 0
                   Bus-Informationen: pci@0000:03:00.0
                   Version: 01
                   Breite: 32 bits
                   Takt: 33MHz
                   Fähigkeiten: pci msi pm pciexpress normal_decode bus_master cap_list
                   Konfiguration: driver=pcieport
                   Ressourcen: irq:36 ioport:3000(Größe=4096) memory:f2700000-f27fffff
                 *-generic
                      Beschreibung: Unassigned class
                      Produkt: RTL8411B PCI Express Card Reader
                      Hersteller: Realtek Semiconductor Co., Ltd.
                      Physische ID: 0
                      Bus-Informationen: pci@0000:04:00.0
                      Version: 01
                      Breite: 32 bits
                      Takt: 33MHz
                      Fähigkeiten: pm msi pciexpress msix vpd bus_master cap_list rom
                      Konfiguration: driver=rtsx_pci latency=0
                      Ressourcen: irq:40 memory:f2705000-f2705fff memory:f2710000-f271ffff
                 *-network
                      Beschreibung: Ethernet interface
                      Produkt: RTL8111/8168/8411 PCI Express Gigabit Ethernet Controller
                      Hersteller: Realtek Semiconductor Co., Ltd.
                      Physische ID: 0.1
                      Bus-Informationen: pci@0000:04:00.1
                      Logischer Name: enp4s0f1
                      Version: 12
                      Seriennummer: 80:fa:5b:8b:a4:43
                      Kapazität: 1Gbit/s
                      Breite: 64 bits
                      Takt: 33MHz
                      Fähigkeiten: pm msi pciexpress msix vpd bus_master cap_list ethernet physical tp 10bt 10bt-fd 100bt 100bt-fd 1000bt-fd autonegotiation
                      Konfiguration: autonegotiation=on broadcast=yes driver=r8168 driverversion=8.048.00-NAPI latency=0 link=no multicast=yes port=twisted pair
                      Ressourcen: irq:46 ioport:3000(Größe=256) memory:f2704000-f2704fff memory:f2700000-f2703fff
              *-pci:1
                   Beschreibung: PCI bridge
                   Produkt: 400 Series Chipset PCIe Port
                   Hersteller: Advanced Micro Devices, Inc. [AMD]
                   Physische ID: 1
                   Bus-Informationen: pci@0000:03:01.0
                   Version: 01
                   Breite: 32 bits
                   Takt: 33MHz
                   Fähigkeiten: pci msi pm pciexpress normal_decode bus_master cap_list
                   Konfiguration: driver=pcieport
                   Ressourcen: irq:38 memory:f2600000-f26fffff
                 *-network
                      Beschreibung: Kabellose Verbindung
                      Produkt: Wi-Fi 6 AX200
                      Hersteller: Intel Corporation
                      Physische ID: 0
                      Bus-Informationen: pci@0000:05:00.0
                      Logischer Name: wlp5s0
                      Version: 1a
                      Seriennummer: c0:3c:59:d0:4a:12
                      Breite: 64 bits
                      Takt: 33MHz
                      Fähigkeiten: pm msi pciexpress msix bus_master cap_list ethernet physical wireless
                      Konfiguration: broadcast=yes driver=iwlwifi driverversion=5.13.0-27-generic firmware=63.c04f3485.0 cc-a0-63.ucode ip=192.168.1.158 latency=0 link=yes multicast=yes wireless=IEEE 802.11
                      Ressourcen: irq:37 memory:f2600000-f2603fff
              *-pci:2
                   Beschreibung: PCI bridge
                   Produkt: 400 Series Chipset PCIe Port
                   Hersteller: Advanced Micro Devices, Inc. [AMD]
                   Physische ID: 4
                   Bus-Informationen: pci@0000:03:04.0
                   Version: 01
                   Breite: 32 bits
                   Takt: 33MHz
                   Fähigkeiten: pci msi pm pciexpress normal_decode bus_master cap_list
                   Konfiguration: driver=pcieport
                   Ressourcen: irq:39
        *-pci:2
             Beschreibung: PCI bridge
             Produkt: Starship/Matisse GPP Bridge
             Hersteller: Advanced Micro Devices, Inc. [AMD]
             Physische ID: 3.1
             Bus-Informationen: pci@0000:00:03.1
             Version: 00
             Breite: 32 bits
             Takt: 33MHz
             Fähigkeiten: pci pm pciexpress msi ht normal_decode bus_master cap_list
             Konfiguration: driver=pcieport
             Ressourcen: irq:28 ioport:2000(Größe=4096) memory:f3000000-f40fffff ioport:e0000000(Größe=303038464)
           *-display
                Beschreibung: VGA compatible controller
                Produkt: NVIDIA Corporation
                Hersteller: NVIDIA Corporation
                Physische ID: 0
                Bus-Informationen: pci@0000:07:00.0
                Version: a1
                Breite: 64 bits
                Takt: 33MHz
                Fähigkeiten: pm msi pciexpress vga_controller bus_master cap_list rom
                Konfiguration: driver=nvidia latency=0
                Ressourcen: irq:129 memory:f3000000-f3ffffff memory:e0000000-efffffff memory:f0000000-f1ffffff ioport:2000(Größe=128) memory:f4080000-f40fffff
           *-multimedia
                Beschreibung: Audio device
                Produkt: TU106 High Definition Audio Controller
                Hersteller: NVIDIA Corporation
                Physische ID: 0.1
                Bus-Informationen: pci@0000:07:00.1
                Version: a1
                Breite: 32 bits
                Takt: 33MHz
                Fähigkeiten: pm msi pciexpress bus_master cap_list
                Konfiguration: driver=snd_hda_intel latency=0
                Ressourcen: irq:125 memory:f4000000-f4003fff
           *-usb
                Beschreibung: USB controller
                Produkt: TU106 USB 3.1 Host Controller
                Hersteller: NVIDIA Corporation
                Physische ID: 0.2
                Bus-Informationen: pci@0000:07:00.2
                Version: a1
                Breite: 64 bits
                Takt: 33MHz
                Fähigkeiten: msi pciexpress pm xhci cap_list
                Konfiguration: driver=xhci_hcd latency=0
                Ressourcen: irq:84 memory:f2000000-f203ffff memory:f2040000-f204ffff
              *-usbhost:0
                   Produkt: xHCI Host Controller
                   Hersteller: Linux 5.13.0-27-generic xhci-hcd
                   Physische ID: 0
                   Bus-Informationen: usb@3
                   Logischer Name: usb3
                   Version: 5.13
                   Fähigkeiten: usb-2.00
                   Konfiguration: driver=hub slots=2 speed=480Mbit/s
              *-usbhost:1
                   Produkt: xHCI Host Controller
                   Hersteller: Linux 5.13.0-27-generic xhci-hcd
                   Physische ID: 1
                   Bus-Informationen: usb@4
                   Logischer Name: usb4
                   Version: 5.13
                   Fähigkeiten: usb-3.10
                   Konfiguration: driver=hub slots=4 speed=10000Mbit/s
           *-serial
                Beschreibung: Serial bus controller
                Produkt: TU106 USB Type-C UCSI Controller
                Hersteller: NVIDIA Corporation
                Physische ID: 0.3
                Bus-Informationen: pci@0000:07:00.3
                Version: a1
                Breite: 32 bits
                Takt: 33MHz
                Fähigkeiten: msi pciexpress pm bus_master cap_list
                Konfiguration: driver=nvidia-gpu latency=0
                Ressourcen: irq:42 memory:f4004000-f4004fff
        *-pci:3
             Beschreibung: PCI bridge
             Produkt: Starship/Matisse Internal PCIe GPP Bridge 0 to bus[E:B]
             Hersteller: Advanced Micro Devices, Inc. [AMD]
             Physische ID: 7.1
             Bus-Informationen: pci@0000:00:07.1
             Version: 00
             Breite: 32 bits
             Takt: 33MHz
             Fähigkeiten: pci pm pciexpress msi ht normal_decode bus_master cap_list
             Konfiguration: driver=pcieport
             Ressourcen: irq:30
           *-generic UNGEFORDERT
                Beschreibung: Non-Essential Instrumentation
                Produkt: Starship/Matisse PCIe Dummy Function
                Hersteller: Advanced Micro Devices, Inc. [AMD]
                Physische ID: 0
                Bus-Informationen: pci@0000:08:00.0
                Version: 00
                Breite: 32 bits
                Takt: 33MHz
                Fähigkeiten: pm pciexpress cap_list
                Konfiguration: latency=0
        *-pci:4
             Beschreibung: PCI bridge
             Produkt: Starship/Matisse Internal PCIe GPP Bridge 0 to bus[E:B]
             Hersteller: Advanced Micro Devices, Inc. [AMD]
             Physische ID: 8.1
             Bus-Informationen: pci@0000:00:08.1
             Version: 00
             Breite: 32 bits
             Takt: 33MHz
             Fähigkeiten: pci pm pciexpress msi ht normal_decode bus_master cap_list
             Konfiguration: driver=pcieport
             Ressourcen: irq:31 memory:f2300000-f25fffff
           *-generic:0 UNGEFORDERT
                Beschreibung: Non-Essential Instrumentation
                Produkt: Starship/Matisse Reserved SPP
                Hersteller: Advanced Micro Devices, Inc. [AMD]
                Physische ID: 0
                Bus-Informationen: pci@0000:09:00.0
                Version: 00
                Breite: 32 bits
                Takt: 33MHz
                Fähigkeiten: pm pciexpress cap_list
                Konfiguration: latency=0
           *-generic:1
                Beschreibung: Encryption controller
                Produkt: Starship/Matisse Cryptographic Coprocessor PSPCPP
                Hersteller: Advanced Micro Devices, Inc. [AMD]
                Physische ID: 0.1
                Bus-Informationen: pci@0000:09:00.1
                Version: 00
                Breite: 32 bits
                Takt: 33MHz
                Fähigkeiten: pm pciexpress msi msix bus_master cap_list
                Konfiguration: driver=ccp latency=0
                Ressourcen: irq:106 memory:f2400000-f24fffff memory:f2508000-f2509fff
           *-usb
                Beschreibung: USB controller
                Produkt: Matisse USB 3.0 Host Controller
                Hersteller: Advanced Micro Devices, Inc. [AMD]
                Physische ID: 0.3
                Bus-Informationen: pci@0000:09:00.3
                Version: 00
                Breite: 64 bits
                Takt: 33MHz
                Fähigkeiten: pm pciexpress msi msix xhci bus_master cap_list
                Konfiguration: driver=xhci_hcd latency=0
                Ressourcen: irq:85 memory:f2300000-f23fffff
              *-usbhost:0
                   Produkt: xHCI Host Controller
                   Hersteller: Linux 5.13.0-27-generic xhci-hcd
                   Physische ID: 0
                   Bus-Informationen: usb@5
                   Logischer Name: usb5
                   Version: 5.13
                   Fähigkeiten: usb-2.00
                   Konfiguration: driver=hub slots=4 speed=480Mbit/s
                 *-usb
                      Beschreibung: Video
                      Produkt: Chicony USB2.0 Camera
                      Hersteller: Sonix Technology Co., Ltd.
                      Physische ID: 4
                      Bus-Informationen: usb@5:4
                      Version: 54.04
                      Fähigkeiten: usb-2.00
                      Konfiguration: driver=uvcvideo maxpower=500mA speed=480Mbit/s
              *-usbhost:1
                   Produkt: xHCI Host Controller
                   Hersteller: Linux 5.13.0-27-generic xhci-hcd
                   Physische ID: 1
                   Bus-Informationen: usb@6
                   Logischer Name: usb6
                   Version: 5.13
                   Fähigkeiten: usb-3.10
                   Konfiguration: driver=hub slots=4 speed=10000Mbit/s
           *-multimedia
                Beschreibung: Audio device
                Produkt: Starship/Matisse HD Audio Controller
                Hersteller: Advanced Micro Devices, Inc. [AMD]
                Physische ID: 0.4
                Bus-Informationen: pci@0000:09:00.4
                Version: 00
                Breite: 32 bits
                Takt: 33MHz
                Fähigkeiten: pm pciexpress msi bus_master cap_list
                Konfiguration: driver=snd_hda_intel latency=0
                Ressourcen: irq:47 memory:f2500000-f2507fff
        *-pci:5
             Beschreibung: PCI bridge
             Produkt: Starship/Matisse Internal PCIe GPP Bridge 0 to bus[E:B]
             Hersteller: Advanced Micro Devices, Inc. [AMD]
             Physische ID: 8.2
             Bus-Informationen: pci@0000:00:08.2
             Version: 00
             Breite: 32 bits
             Takt: 33MHz
             Fähigkeiten: pci pm pciexpress msi ht normal_decode bus_master cap_list
             Konfiguration: driver=pcieport
             Ressourcen: irq:32 memory:f2200000-f22fffff
           *-sata
                Beschreibung: SATA controller
                Produkt: FCH SATA Controller [AHCI mode]
                Hersteller: Advanced Micro Devices, Inc. [AMD]
                Physische ID: 0
                Bus-Informationen: pci@0000:0a:00.0
                Version: 51
                Breite: 32 bits
                Takt: 33MHz
                Fähigkeiten: sata pm pciexpress msi ahci_1.0 bus_master cap_list
                Konfiguration: driver=ahci latency=0
                Ressourcen: irq:49 memory:f2200000-f22007ff
        *-pci:6
             Beschreibung: PCI bridge
             Produkt: Starship/Matisse Internal PCIe GPP Bridge 0 to bus[E:B]
             Hersteller: Advanced Micro Devices, Inc. [AMD]
             Physische ID: 8.3
             Bus-Informationen: pci@0000:00:08.3
             Version: 00
             Breite: 32 bits
             Takt: 33MHz
             Fähigkeiten: pci pm pciexpress msi ht normal_decode bus_master cap_list
             Konfiguration: driver=pcieport
             Ressourcen: irq:33 memory:f2100000-f21fffff
           *-sata
                Beschreibung: SATA controller
                Produkt: FCH SATA Controller [AHCI mode]
                Hersteller: Advanced Micro Devices, Inc. [AMD]
                Physische ID: 0
                Bus-Informationen: pci@0000:0b:00.0
                Version: 51
                Breite: 32 bits
                Takt: 33MHz
                Fähigkeiten: sata pm pciexpress msi ahci_1.0 bus_master cap_list
                Konfiguration: driver=ahci latency=0
                Ressourcen: irq:66 memory:f2100000-f21007ff
        *-serial
             Beschreibung: SMBus
             Produkt: FCH SMBus Controller
             Hersteller: Advanced Micro Devices, Inc. [AMD]
             Physische ID: 14
             Bus-Informationen: pci@0000:00:14.0
             Version: 61
             Breite: 32 bits
             Takt: 66MHz
             Konfiguration: driver=piix4_smbus latency=0
             Ressourcen: irq:0
        *-isa
             Beschreibung: ISA bridge
             Produkt: FCH LPC Bridge
             Hersteller: Advanced Micro Devices, Inc. [AMD]
             Physische ID: 14.3
             Bus-Informationen: pci@0000:00:14.3
             Version: 51
             Breite: 32 bits
             Takt: 66MHz
             Fähigkeiten: isa bus_master
             Konfiguration: latency=0
     *-pci:1
          Beschreibung: Host bridge
          Produkt: Starship/Matisse PCIe Dummy Host Bridge
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 101
          Bus-Informationen: pci@0000:00:01.0
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
     *-pci:2
          Beschreibung: Host bridge
          Produkt: Starship/Matisse PCIe Dummy Host Bridge
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 102
          Bus-Informationen: pci@0000:00:02.0
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
     *-pci:3
          Beschreibung: Host bridge
          Produkt: Starship/Matisse PCIe Dummy Host Bridge
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 103
          Bus-Informationen: pci@0000:00:03.0
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
     *-pci:4
          Beschreibung: Host bridge
          Produkt: Starship/Matisse PCIe Dummy Host Bridge
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 104
          Bus-Informationen: pci@0000:00:04.0
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
     *-pci:5
          Beschreibung: Host bridge
          Produkt: Starship/Matisse PCIe Dummy Host Bridge
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 105
          Bus-Informationen: pci@0000:00:05.0
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
     *-pci:6
          Beschreibung: Host bridge
          Produkt: Starship/Matisse PCIe Dummy Host Bridge
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 106
          Bus-Informationen: pci@0000:00:07.0
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
     *-pci:7
          Beschreibung: Host bridge
          Produkt: Starship/Matisse PCIe Dummy Host Bridge
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 107
          Bus-Informationen: pci@0000:00:08.0
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
     *-pci:8
          Beschreibung: Host bridge
          Produkt: Matisse Device 24: Function 0
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 108
          Bus-Informationen: pci@0000:00:18.0
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
     *-pci:9
          Beschreibung: Host bridge
          Produkt: Matisse Device 24: Function 1
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 109
          Bus-Informationen: pci@0000:00:18.1
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
     *-pci:10
          Beschreibung: Host bridge
          Produkt: Matisse Device 24: Function 2
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 10a
          Bus-Informationen: pci@0000:00:18.2
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
     *-pci:11
          Beschreibung: Host bridge
          Produkt: Matisse Device 24: Function 3
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 10b
          Bus-Informationen: pci@0000:00:18.3
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
          Konfiguration: driver=k10temp
          Ressourcen: irq:0
     *-pci:12
          Beschreibung: Host bridge
          Produkt: Matisse Device 24: Function 4
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 10c
          Bus-Informationen: pci@0000:00:18.4
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
     *-pci:13
          Beschreibung: Host bridge
          Produkt: Matisse Device 24: Function 5
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 10d
          Bus-Informationen: pci@0000:00:18.5
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
     *-pci:14
          Beschreibung: Host bridge
          Produkt: Matisse Device 24: Function 6
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 10e
          Bus-Informationen: pci@0000:00:18.6
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
     *-pci:15
          Beschreibung: Host bridge
          Produkt: Matisse Device 24: Function 7
          Hersteller: Advanced Micro Devices, Inc. [AMD]
          Physische ID: 10f
          Bus-Informationen: pci@0000:00:18.7
          Version: 00
          Breite: 32 bits
          Takt: 33MHz
     *-pnp00:00
          Produkt: PnP device PNP0c02
          Physische ID: 1
          Fähigkeiten: pnp
          Konfiguration: driver=system
     *-pnp00:01
          Produkt: PnP device PNP0b00
          Physische ID: 2
          Fähigkeiten: pnp
          Konfiguration: driver=rtc_cmos
     *-pnp00:02
          Produkt: PnP device PNP0c02
          Physische ID: 3
          Fähigkeiten: pnp
          Konfiguration: driver=system
     *-pnp00:03
          Produkt: PnP device PNP0c01
          Physische ID: 5
          Fähigkeiten: pnp
          Konfiguration: driver=system
     *-pnp00:04
          Produkt: PnP device PNP0303
          Physische ID: 6
          Fähigkeiten: pnp
          Konfiguration: driver=i8042 kbd
     *-pnp00:05
          Produkt: PnP device PNP0f13
          Physische ID: 7
          Fähigkeiten: pnp
          Konfiguration: driver=i8042 aux



Ubuntu 20.04 LTS

using Rosetta project and SiDock project


Any workaround so far?

**Comment by JoesCat** [comment]
Hi - has anyone looked at running 'nice' in regards to boinc?
If you have a sufficiently high 'nice' level, let's say 19, your computer CPU(s) can be running 100%, and yet the computer does not feel sluggish at all. Watch a movie, do other number crunching, seems to work fine. Example: http://www.joescat.com/boinc/#PRJ
```
[boinc@genesis ~]$ top
  PID USER   PR NI  VIRT  RES  SHR S %CPU %MEM    TIME+  COMMAND          
 7287 boinc  34 19 62440  49m 2236 S 98.7  6.8 163:53.72 einstien_S5R3_4
 3457 root   15  0 57524  16m 4192 S  0.7  2.3   4:52.38 X
 3300 boinc  26 10  5060 3138 1748 S  0.3  0.4   0:18:34 boinc
```

**Comment by CubeTheThird** [comment]
> Hi - has anyone looked at running 'nice' in regards to boinc? If you have a sufficiently high 'nice' level, let's say 19, your computer CPU(s) can be running 100%, and yet the computer does not feel sluggish at all. Watch a movie, do other number crunching, seems to work fine. Example: http://www.joescat.com/boinc/#PRJ

Although this _can_ be helpful, unfortunately nice levels apply to CPU scheduling. If you are doing GPU computations, I doubt changing the nice level will have impact on system performance.

**Comment by siriak** [comment]
This is unfortunate that idle detection doesn't work. This prevents me and I imaging a lot of other people from running BOINC.

**Comment by makeasnek** [comment]
A [bounty](https://app.bountysource.com/issues/110844489-boinc-bug-bounty-fix-linux-idle-detection) for this issue has been started by the the [SCI](https://thesciencecommons.org). If this issue is important to you, consider contributing to this bounty. The higher the bounty, the greater the chance a developer outside of BOINC will see it and fix it. The bounty is currently at $100 USD. You can contribute to the bounty, check the total amount pledged, or claim the bounty at https://app.bountysource.com/issues/110844489-boinc-bug-bounty-fix-linux-idle-detection

**Comment by BryanQuigley** [comment]
just found https://wayland.app/protocols/ext-idle-notify-v1 (and ChatGPT actually can just write with it because I couldn't find an example... ) but obviously it has the same issue as the BOINC service won't be able to access the wayland session. 

I see 3 paths to "fix" this:
1) Accept BOINC needs to run as a Flatpak or similar user app and don't run BOINC as a system service at all.  Is annoying that it won't run if the App isn't open and has less security from your session.
2) A user space utility that controls (or communicates with) the system BOINC service as to how idle you are. (Maybe something like https://github.com/swaywm/swayidle/blob/master/swayidle.1.scd could do it for sway/KDE?).  This will be desktop environment (and X11 vs Wayland) dependent.
3) Just get better CPU/GPU usage detection to determine if BOINC is impacting other processes. It was almost there for me last time I used BOINC just by setting "In use" to be 5% CPU.  Improving this might be video card specific (nvidia might use nvidia-smi).  It may also be easier to do an external utility that controls BOINC even in this case.

Hope these ideas help someone looking at this

**Comment by siriak** [comment]
@BryanQuigley check out the script I wrote which controls how much CPU BOINC is using. https://github.com/siriak/boinc-load-balancer
Maybe you or somebody else could do something similar for managing GPU load.

**Comment by LinAGKar** [comment]
Maybe it would be better if the idle detection was handled by boinc-manager rather than boinc-client, since that runs in the user's session. Ideally then, boinc-manager should get tray support.

**Comment by AenBleidd** [comment]
> Maybe it would be better if the idle detection was handled by boinc-manager rather than boinc-client, since that runs in the user's session. 

What is the user runs no BOINC Manager on that particular machine?

**Comment by makeasnek** [comment]
It seems the solution would be to have BOINC client accept multiple idle detection types, some of which the client can access directly, some of which exist in userspace (ie BOINC manager) which can be communicated to the client via RPC. And then the client should pick which of these signals to trust based on some internal logic.

**Comment by CharlieFenton** [comment]
What is the user runs no BOINC Manager on that particular machine?

> While I.m not as familiar with Linux as others, I want to add that it might be possible for the client to launch a small helper application in user space independent of BOINC Manager.

**Comment by makeasnek** [comment]
@CharlieFenton I was under the impression, perhaps incorrectly, that BOINC already detects correct idle state in these machines by monitoring keyboard/mouse input on some /dev device.

**Comment by CharlieFenton** [comment]
@makeasnek You very well may be correct; I don't know. As I mentioned, I am not as family with Linux and don't have a Linux box. I was just responding to [this commen](https://github.com/BOINC/boinc/issues/1187#issuecomment-1532269440)t by @LinAGKar and [this response](https://github.com/BOINC/boinc/issues/1187#issuecomment-1531671796) from @AenBleidd.

**Comment by RichardHaselgrove** [comment]
And never forget the case where the user is 'using' the device, with full concentration - but not actively making any inputs. Watching films full-screen, for instance.

**Comment by 0x0OZ** [comment]
Where is the broken idle check stored in the repo?

**Comment by AenBleidd** [comment]
> Where is the broken idle check in the repo located? 

It's not broken, rather it need to be redefined in a proper way.
The code is here: https://github.com/BOINC/boinc/blob/eac3e3a2d9fc74b95290273ebe2e9b0b9b133997/client/hostinfo_unix.cpp
Please keep in mind that for different OSs we have different idle detection mechanisms.

**Comment by 0x0OZ** [comment]
> > Where is the broken idle check in the repo located?
> 
> It's not broken, rather it need to be redefined in a proper way. The code is here: https://github.com/BOINC/boinc/blob/master/client/hostinfo_unix.cpp Please keep in mind that for different OSs we have different idle detection mechanisms.

I see it's defined for Xorg, but it's not working and not defined for wayland.

**Comment by AenBleidd** [comment]
> > > Where is the broken idle check in the repo located?
> > 
> > 
> > It's not broken, rather it need to be redefined in a proper way. The code is here: https://github.com/BOINC/boinc/blob/master/client/hostinfo_unix.cpp Please keep in mind that for different OSs we have different idle detection mechanisms.
> 
> I see it's defined for Xorg, but it's not working and not defined for wayland.

Because it was created when there was no Wayland.
That is why I said that it's not broken.
It's just need to be properly implemented for Wayland (ok, there're also some issues with the XOrg, mostly for headless devices, but still).

**Comment by davidpanderson** [comment]
My recollection is that X-based idle detection doesn't work
because the BOINC client runs as a different user,
and that we need to use an approach like 'BOINC system tray' on Windows.
I never got around to implementing this, but could do so.

**Comment by 0x0OZ** [comment]
> My recollection is that X-based idle detection doesn't work because the BOINC client runs as a different user, and that we need to use an approach like 'BOINC system tray' on Windows. I never got around to implementing this, but could do so.

That seems right, I couldn't run `xprintidle` on a different user...
Having a system tray won't be a global solution as you have headless systems that use boinc as mentioned by @AenBleidd .

I think a global solution for all cases would be detecting if there is any user input through mouse/keyboard and if there is any output sound for *X* seconds to decide that which I don't know how. 
Also in the case of Xorg built-in idle detection, it does not count ssh connections as activity, which means if you use an ssh to connect to a distro with Xorg, the system will be IDLE.

**Comment by RichardHaselgrove** [comment]
There's more discussion of this in #4125 - a specific problem with enumerating X-servers under systemd.

**Comment by 0x0OZ** [comment]
I made a small patch that _should_  fix the issue globally for all linux distributions for different cases (headless / xorg /wayland).
But because cpp is not really my language I will not be able to make the code clean or add it in the way it should be implemented in the repo...

I only added the method to be reviewed, I don't want to break the code by removing methods like `xss_idle()`

**Comment by Drag-NDrop** [comment]
For inspiration; I've created a prototype to help with headless devices.
The main thougt is, that no matter how one connect to a headless instance, one would be using a terminal. So, IdleRunner was made for Bash. Wouldn't be to hard to make some conditional logic that took other shells into account.

**Comment by Sandman192** [comment]
This was reported with 7.0.28 of 2015 and BOINC 8.0.2 still has a problem with not detecting mouse or keyboard to suspend work. But with Mint 22.1 for me.
It seems to have nothing to do with what Linux distribution it is. But with BOINC it's self.
Using `sudo apt install boinc-client boinc-manager`

Suspend debug every second `Tue 01 Jul 2025 09:57:12 PM CDT |  | [suspend] net_susp: no; file_xfer_susp: no; reason: unknown reason`

**Comment by nicepopo86-lang** [comment]
I’m starting investigation and implementation for Linux idle detection reliability (including Wayland/X paths and a helper-process approach). I’ll share a concrete patch proposal and test results on Ubuntu regular + LTS.

**Comment by AenBleidd** [comment]
@nicepopo86-lang, this is not needed since this is already done: https://github.com/jamescowens/idle_detect
The integration [PR](#6242) is open and is just waiting to be properly tested and merged.

**Comment by nicepopo86-lang** [comment]
Phase-1 patch is up as a draft PR: https://github.com/BOINC/boinc/pull/6865

Delivered in this phase:
- optional Linux helper RPC probe over UNIX socket (`GET_IDLE_SECS\n` -> parse `OK <seconds> <source>`)
- integration into `HOST_INFO::user_idle_time()` before existing XSS fallback
- safe fallback to existing detection path when helper is disabled/unavailable
- new minimal `cc_config.xml` knobs: `enable_idle_helper_rpc`, `idle_helper_socket`, `idle_helper_timeout_ms` (defaults keep current behavior)

Note: full local compile could not be completed in this environment because `./_autosetup` fails due to missing `m4`.
